### PR TITLE
executor can use a different sparkHome from Worker 

### PR DESCRIPTION
**Pull request by squito**
We'd really like to be able to specify a different sparkHome for an exector, without having to restart the Workers on the cluster.  This would really make it much easier for us to test out small changes for spark -- we'd like to simply leave the Master/Workers running on the cluster, and be able to try out different versions of spark for the executors by setting sparkHome.  We really don't want to just replace the contents of the sparkHome dir, b/c we'd like to have multiple versions around at the same time -- there may be other things using the cluster at the same time as I test out spark versions.

I realize this isn't foolproof -- eventually there may be incompatible versions of the Worker & Executor.  But I'd still like to be able to delay restarting the cluster as much as I can.

I am not sure how this relates to the mesos cluster modes ...

(Without this change, isn't the sparkHome arg to SparkContext completely ignored in standalone clusters?)

### CODE REVIEW

File: core/src/main/scala/spark/deploy/DeployMessage.scala [hunk 1]
```diff
@@ -41,7 +42,8 @@ private[spark] case class LaunchExecutor(
     execId: Int,
     jobDesc: JobDescription,
     cores: Int,
-    memory: Int)
+    memory: Int,
+    sparkHome: File)
```
Comment (mateiz) [inline core/src/main/scala/spark/deploy/DeployMessage.scala]: I'd make this a String -- while File is Serializable, it looks a little weird to send it across the network.

**Comment by mateiz** [comment]
Thanks Imran. I'm not sure this is something we'll be able to support in the long run (certainly not across Spark versions for example), but since it's useful to you I think it's fine to include it. I made one comment on the code though.

**Comment by squito** [comment]
I've updated the code to send the dir as a String.

I definitely understand that arbitrary Worker / StandaloneExecutorBackends won't work together.  But I'm just curious -- does the sparkHome arg to SparkContext have any other effect?

thanks

File: core/src/main/scala/spark/deploy/JobDescription.scala [hunk 2]
```diff
@@ -1,10 +1,13 @@
...
 
+import java.io.File
+
 private[spark] class JobDescription(
     val name: String,
     val cores: Int,
     val memoryPerSlave: Int,
-    val command: Command)
+    val command: Command,
+    val sparkHome: File)
```
Comment (mateiz) [inline core/src/main/scala/spark/deploy/JobDescription.scala]: Sorry, this should also be a String for the same reason as above

**Comment by squito** [comment]
oops, sorry I should have caught that.  code updated.

**Comment by mateiz** [comment]
Great, thanks!

**Comment by mateiz** [comment]
I've cherry-picked this into master as well BTW.

**Pull request merged**